contrib/segmentio/kafka.go.v0/kafka.go

```diff
@@ -0,0 +1,149 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package kafka
+
+import (
+	"context"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+
+	"github.com/segmentio/kafka-go"
+)
+
+// NewReader calls kafka.NewReader and wraps the resulting Consumer.
+func NewReader(conf kafka.ReaderConfig, opts ...Option) *Reader {
+	return WrapReader(kafka.NewReader(conf), opts...)
+}
+
+// NewWriter calls kafka.NewWriter and wraps the resulting Producer.
+func NewWriter(conf kafka.WriterConfig, opts ...Option) *Writer {
+	return WrapWriter(kafka.NewWriter(conf), opts...)
+}
+
+// WrapReader wraps a kafka.Reader so that any consumed events are traced.
+func WrapReader(c *kafka.Reader, opts ...Option) *Reader {
+	wrapped := &Reader{
+		Reader: c,
+		cfg:    newConfig(opts...),
+	}
+	log.Debug("contrib/confluentinc/confluent-kafka.go.v0/kafka: Wrapping Reader: %#v", wrapped.cfg)
+	return wrapped
+}
+
+// A Reader wraps a kafka.Reader.
+type Reader struct {
+	*kafka.Reader
+	cfg  *config
+	prev ddtrace.Span
+}
+
+func (r *Reader) startSpan(ctx context.Context, msg *kafka.Message) ddtrace.Span {
+	opts := []tracer.StartSpanOption{
+		tracer.ServiceName(r.cfg.consumerServiceName),
+		tracer.ResourceName("Consume Topic " + msg.Topic),
+		tracer.SpanType(ext.SpanTypeMessageConsumer),
+		tracer.Tag("partition", msg.Partition),
+		tracer.Tag("offset", msg.Offset),
+		tracer.Measured(),
+	}
+	if !math.IsNaN(r.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, r.cfg.analyticsRate))
+	}
+	// kafka supports headers, so try to extract a span context
+	carrier := MessageCarrier{msg}
+	if spanctx, err := tracer.Extract(carrier); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+	span, _ := tracer.StartSpanFromContext(ctx, "kafka.consume", opts...)
+	// reinject the span context so consumers can pick it up
+	if err := tracer.Inject(span.Context(), carrier); err != nil {
+		log.Debug("contrib/segmentio/kafka.go.v0: Failed to inject span context into carrier, %v", err)
+	}
+	return span
+}
+
+// Close calls the underlying Reader.Close and if polling is enabled, finishes
+// any remaining span.
+func (r *Reader) Close() error {
+	err := r.Reader.Close()
+	if r.prev != nil {
+		r.prev.Finish()
+		r.prev = nil
+	}
+	return err
+}
+
+// ReadMessage polls the consumer for a message. Message will be traced.
+func (r *Reader) ReadMessage(ctx context.Context) (kafka.Message, error) {
+	if r.prev != nil {
+		r.prev.Finish()
+		r.prev = nil
+	}
+	msg, err := r.Reader.ReadMessage(ctx)
+	if err != nil {
+		return kafka.Message{}, err
+	}
+	r.prev = r.startSpan(ctx, &msg)
+	return msg, nil
+}
```

Comment:
I don't have any suggestions here, just thinking out loud...
As this is written, the only way a span gets closed is when ReadMessage() is called again, or when Close() is called. 

This is probably approximately accurate when messages are handled in a loop, i.e.:
```go
for m, err = r.ReadMessage(...); err == nil {
    handleMessage(m)
}
```

Even if `handleMessage(m)` handles things asynchronously (meaning the span from ReadMessage ends while the message is still being handled), the trace can still continue since `m` carries the `SpanContext`. The user will just have to start a new span.
```go
for m, err = r.ReadMessage(...); err == nil {
    go handleMessage(m)
    // handleMessage is running, r.ReadMessage() is called and ends the `kafka.consume` span
}
```

This is a little funny, since as long as the `Reader` is open and *not* in an active call to `ReadMessage`, there will be an active span. 

Does this seem OK? Are there use cases where this doesn't work well? Do users write kafka consumers some way other than a receive loop?

Reply:
This is a good point, I wasn't looking too closely at exactly how the spans were matching up with the operations actually happening here from the previous PR. This code was originally heavily based on our pre-existing `confluentinc/confluent-kafka-go` integration which seems to do almost the same thing.

Looking closer at that integration it seems to have nearly the same "issue".  Although I think it's more clear that this new integration shouldn't do spans this way. The docs for the segmentio `Reader` do heavily encourage using the `Close` method, but with the way spans are created here it seems like all the tracing is "offset" from what should really be getting traced. If I understand correctly it looks a bit like this:
```
ReadMessage
	segmentio.reader.ReadMessage BLOCKS
	startSpan
User Processes message
Close
	finishSpan
```
when what we probably want is closer to:
```
ReadMessage
    startSpan
    segmentio.reader.ReadMessage BLOCKS
    finishSpan
User Processes message
Close
```
Any thoughts on if this makes more sense for this integration? Additionally, we may want to also make this change to the confluent-kafka-go integration to more accurately trace kafka operations.

**Even more information that I'm writing out for my own benefit and is definitely not important for understanding my comment above:**
Looking closely at the third integration `shopify/sarama`. I think that the above issues are copy-paste problems that originated in trying to follow the pattern set there. `sarama` handles message consumption in an async way, so the method used for tracing there involves starting spans after having seen a message and finishing the span after the next message is received. It also feels a bit "off by one" in nature, but I think only really matters for the very first message that is consumed where the time spent connecting / receiving the message from kafka isn't included in the span, but the time spent getting the next message from the underlying `sarama.Messages` channel will be. This slight "offset" seems to be an intentional choice to reduce complexity in tracing the messages as they come though a channel and must be sent out via another channel to the client of the library.

Reply:
I think that customers *do* want to measure the `User Processes Message` section of the above list. That's likely the most important thing to measure - how long the user code takes to process a message.

Ideally, this could be similar to http request tracing
```
Request Received
Start Span
User Processes Message
End Span
Return Response
```

Kafka doesn't have this request/response model, though. It's one-way, receive-only.
Knowing how much time is spent in `segmentio.reader.ReadMessage BLOCKS` may be useful, but it would definitely be a case of long running spans. Sometimes messages might come in very rapidly. Other times there may be many minutes between messages.

We also don't want the BLOCKS time to overshadow the actual application logic time. Not sure how to solve this one *including* blocking time. 

The API doesn't give us a guaranteed way to know when processing starts and finishes. We know when a message gets handed to the user, but don't know when they finish - we just know when they call `ReadMessage` next. There's no [Message](https://pkg.go.dev/github.com/segmentio/kafka-go#Message).Finish() or something like that.

We *might* be able to provide a function like:
```
func TraceMessage(ctx context.Context, k kafka.Message, handler func()) {
    span, ctx := tracer.StartSpanFromContext(ctx, ...)
    defer span.Finish()
    handler()
}
```

Then the user could write handlers like this:
```
for m, err = r.ReadMessage(...); err == nil {
    kafkatrace.TraceMessage(ctx, m, func() {
        handleMessage(m)
    })
}
```

But that's kind of cumbersome and I don't know if users really want to do that.

Reply:
After chatting, I think we can close this. Probably can't do better today and this is how the others work.